### DISCUSSION TITLE
Bad Request : "PrismaClientValidationError"

### DISCUSSION THREAD

**Discussion by IAmMythoRED** [prisma/prisma#21669] [Q&A] [answered]
### Question

We've recently made updates to our code schema, which looks something like this:
```prisma
model Code {
  userId                String      @unique
  code                  String      @unique
  isActivated           Boolean     
  createdAt             DateTime    @default(now())
}
```
In our API code, we update the code or create a new code with the 'isActivated' flag set to 'false.'

Here's the code snippet:
```js
try {
  const codeData = await Prisma.code.findUnique({
    where: {
      userId: userId,
    },
  });

  if (codeData) {
    const updateUser = await Prisma.code.update({
      where: {
        userId: userId,
      },
      data: {
        code: code,
        isActivated: isActivated,
      },
    });
  } else {
    const newCode = await Prisma.code.create({
      data: {
        userId: userId,
        code: code,
        isActivated: isActivated,
      },
    });
  }

  return res.status(200).json({ message: "Generated Code!", code: code });
} catch (error) {
  return res.status(500).json(error);
}
```
We've tested this code with Postman and via our localhost web server.

In Postman, we received the following error:
```
{
    "name": "PrismaClientValidationError",
    "clientVersion": "5.5.2"
}
```
On the web client, the error appeared as:
```
xhr.js:256     POST http://localhost:3000/api/auth/login/code 500 (Internal Server Error)
```
### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}


model Bot {
  id                String      @id @default(cuid())
  botId             String      
  botName           String
  ownerId           String      
  owner             User        @relation(fields: [ownerId],references: [userId])
}
     
model User {
  id                String      @id @default(cuid())
  userId            String      @unique
  userName          String  
  profilePicture    String 
  createdAt         DateTime    @default(now())
  updatedAt         DateTime    @default(now()) @updatedAt
  bots              Bot[]      
  sessionToken      String?
}

model Code {
  userId                String      @unique
  code                  String      @unique
  isActivated           Boolean     
  createdAt             DateTime    @default(now())
}
```

**Comment by ludralph** [comment]
Hi @BotBuddies 👋 

Thank you for raising this question.

The [error](https://www.prisma.io/docs/reference/api-reference/error-reference#prismaclientvalidationerror) message suggests that you have a failing validation. For example, 

- Missing field - for example, an empty data: {} property when creating a new record.
- Incorrect field type provided (for example, setting a Boolean field to "Hello, I like cheese and gold!").

It indicates a problem with the data you're trying to send to the database. Ensure that the data you are sending in the request is valid and matches the schema of your `Code` model.  Add some [logging](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#log-to-stdout) to your code to see the exact values of `userId`, `code`, and `isActivated` that are being used when calling Prisma methods. This will help you identify if there are any unexpected values causing the validation error.

Also, it appears you are trying to [update or create a record](https://www.prisma.io/docs/concepts/components/prisma-client/crud#update-or-create-records) which looks like you can leverage `upsert` method in Prisma.

You can further simplify and improve your code by using the `upsert` operation provided by Prisma. `upsert` is designed for scenarios where you want to either insert a new record or update an existing one based on a certain condition. Here's how you can use upsert to improve your code:
```js
try {
  // Use Prisma upsert to insert or update the record
  const result = await Prisma.code.upsert({
    where: { userId },
    update: {
      code,
      isActivated,
    },
    create: {
      userId,
      code,
      isActivated,
    },
  });

  return res.status(200).json({ message: "Generated Code!", code });
} catch (error) {
  console.error(error); // Log the error for debugging
  return res.status(500).json({ error: "Internal Server Error" });
}
```
If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by IAmMythoRED** [comment]
Okay i understood that the error suggest that some fields are missing when updating or crating, but my question where is it? i can see no issues in my current code

```js
 try {
      const codeData = await Prisma.code.findUnique({
        where: {
          userId: new_userId,
        },
      });

      if (codeData) {
        const updateUser = await Prisma.code.update({
          where: {
            userId: new_userId,
          },
          data: {
            code: new_code,
            isActivated: false
          },
        });
      } else {
        const newCode = await Prisma.code.create({
          data: {
            userId: userId,
            code: code,
            isActivated: false
           },
        });
      }
      ```
      
      My prisma schema is:
      
      // This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Bot {
  id String @id @default(cuid())
  botId String
  botName String
  ownerId String
  owner User @relation(fields: [ownerId],references: [userId])
}

model User {
  id String @id @default(cuid())
  userId String @unique
  userName String
  profilePicture String
  createdAt DateTime @default(now())
  updatedAt DateTime @default(now()) @updatedAt
  bots Bot[]
  sessionToken String?
}

model Code {
  userId String @unique
  code String @unique
  isActivated Boolean @default(false)
  createdAt DateTime @default(now())
}


on code schema u can see i have isActivated and i m filling the isActivated on create and update but it returns prisma client error. This happens whenever i add isActivated datatype.

I hope someone could help on fixing this issue.

**Comment by IAmMythoRED** [comment, marked as answer]
Okay the issue was somehow the import of Prisma client was wrong! we changed the import and it got fixed! Anyway thankyou and we will use the upsert method for updating r creating new data
